### DISCUSSION TITLE
Is there any notable difference between completions from files vs directly sourced?

### DISCUSSION THREAD

**Discussion by flokoe** [fish-shell/fish-shell#10826] [Q&A] [answered]
Hi there,

Many CLI tools provide completions as files or directly via CLI. For example, `just --completions fish`. The content is the same, but is there a recommended way to use completions? Any advantages using the files via `~/.config/fish/completions/just.fish` vs. actively sourcing the completions in the config like this:

```fish
if status is-interactive
    just --completions fish | source
end
```

**Comment by faho** [comment, marked as answer]
The file is only loaded when necessary, i.e. when you do `just <TAB>`.

If you put it in your config like that, it will always run in every interactive shell even if you never use just there.

For just `just` that's probably okay (takes ~3ms on my system), but do that with more or slower tools and it will slow down your shell startup.

**Reply by faho** [reply]
E.g. `deno` takes 20ms, `pip` takes 150ms. My normal fish config takes 10ms to start, so either of those would baloon it, and pip would *definitely* be noticeable.

**Reply by flokoe** [reply]
I see, thanks!
